### PR TITLE
feat(ampd): multisig contract signing started handler

### PR DESCRIPTION
**Pull request by haiyizxx**
## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: ampd/src/handlers/multisig.rs [hunk 1]
```diff
@@ -72,21 +79,126 @@ impl TryFrom<&event_sub::Event> for Option<SigningStartedEvent> {
...
+{
+    type Err = Error;
+
+    async fn handle(&self, event: &event_sub::Event) -> error_stack::Result<(), Error> {
+        let SigningStartedEvent {
+            contract_address,
+            session_id,
+            pub_keys,
+            msg,
+        } = match event.try_into().into_report()? {
```
Comment (sammy1991106) [inline ampd/src/handlers/multisig.rs:148]: use the macro that Chris created.

File: ampd/src/handlers/multisig.rs [hunk 2]
```diff
@@ -72,21 +79,126 @@ impl TryFrom<&event_sub::Event> for Option<SigningStartedEvent> {
...
+                msg,
+                &pub_keys[&self.worker],
+            )
+            .await
+            .change_context(Error::Sign)?;
+
+        self.broadcast_signature(session_id.into(), signature.into())
+            .await?;
+
+        return Ok(());
```
Comment (sammy1991106) [inline ampd/src/handlers/multisig.rs:174]:
```suggestion
        match pub_keys.get(&self.worker) {
            Some(pub_key) => {
                let signature = self
                    .signer
                    .sign(self.multisig.to_string().as_str(), msg, &pub_key)
                    .await
                    .change_context(Error::Sign)?;

                self.broadcast_signature(session_id.into(), signature.into())
                    .await?;

                Ok(())
            }
            None => Ok(()),
        }
```

File: ampd/src/handlers/multisig.rs [hunk 3]
```diff
@@ -72,21 +79,126 @@ impl TryFrom<&event_sub::Event> for Option<SigningStartedEvent> {
...
+
+    async fn broadcast_signature(
+        &self,
+        session_id: Uint64,
+        signature: HexBinary,
+    ) -> error_stack::Result<(), Error> {
+        let msg = serde_json::to_vec(&ExecuteMsg::SubmitSignature {
+            session_id,
+            signature,
+        })
```
Comment (sammy1991106) [inline ampd/src/handlers/multisig.rs:118]:
```suggestion
    async fn broadcast_signature(
        &self,
        session_id: impl Into<Uint64>,
        signature: impl Into<HexBinary>,
    ) -> error_stack::Result<(), Error> {
        let msg = serde_json::to_vec(&ExecuteMsg::SubmitSignature {
            session_id: session_id.into(),
            signature: signature.into(),
        })
```

**Review by sammy1991106 — CHANGES_REQUESTED** [review]

File: ampd/src/handlers/multisig.rs (hunk 1 above)
Comment (haiyizxx) [reply ampd/src/handlers/multisig.rs:148]: was not merged before, will update

File: ampd/src/config.rs [hunk 4]
```diff
@@ -18,6 +19,7 @@ pub struct Config {
     #[serde(with = "hex")]
     pub private_key: ECDSASigningKey,
     pub event_buffer_cap: usize,
+    pub multisig_contract: Option<TMAddress>,
```
[the remark this answers is not in the document]
Comment (haiyizxx) [reply ampd/src/config.rs:22]: We impl Default for config, there is no default value for a contract address. 
It also allows someone run apmd if he wants to use some other proofing mechanism.

File: ampd/src/handlers/multisig.rs [hunk 5]
```diff
@@ -52,23 +60,126 @@ where
...
+        } = match event.try_into() as error_stack::Result<_, _> {
+            Err(report) if matches!(report.current_context(), EventTypeMismatch(_)) => {
+                return Ok(());
+            }
+            result => result.change_context(DeserializeEvent)?,
+        };
+
+        if self.multisig != contract_address {
+            return Ok(());
+        }
```
Comment (cjcobb23) [inline ampd/src/handlers/multisig.rs:138]: I think we need to check the prover address that triggered this signing somehow. I don't think we should sign any data submitted to the multisig contract without knowing where that data came from and what it is.

**Review by cjcobb23 — CHANGES_REQUESTED** [review]

File: ampd/src/handlers/multisig.rs (hunk 5 above)
Comment (haiyizxx) [reply ampd/src/handlers/multisig.rs:138]: if we don't want permissionless signing, we should restrict on the smart contract level, not in ampd

Comment (cjcobb23) [reply ampd/src/handlers/multisig.rs:138]: as in restrict which contracts can trigger signing sessions? We could require contracts to be approved via governance vote before they can trigger signing sessions. We'd need to add a new execute msg to the multisig contract to authorize contracts.

File: ampd/src/handlers/multisig.rs [hunk 6]
```diff
@@ -52,23 +60,126 @@ where
...
+
```
Comment (cjcobb23) [inline ampd/src/handlers/multisig.rs:139]: Can we add a logline here? Just something indicating a signing started event was received. Ampd was quite silent last time I ran it, and it was pretty hard to figure out if it was working or not

File: ampd/src/handlers/multisig.rs [hunk 7]
```diff
@@ -52,23 +60,126 @@ where
...
+                    .sign(self.multisig.to_string().as_str(), msg, pub_key)
+                    .await
+                    .change_context(Error::Sign)?;
+
+                self.broadcast_signature(session_id.into(), signature.into())
+                    .await?;
+
+                Ok(())
+            }
+            None => Ok(()),
```
Comment (cjcobb23) [inline ampd/src/handlers/multisig.rs]: might be useful to add some logging here too, possibly in both match cases

**Review by sammy1991106 — APPROVED** [review]

**Review by cgorenflo — APPROVED** [review]

**Review by cjcobb23 — APPROVED** [review]

**Pull request merged**